igraph_colored_simple_graph

szhorvat opened the pull request:
Currently, no included isomorphism algorithm supports multigraphs.

VF2 does not seem to support either self loops (#1166, please verify) or multigraphs (#815)

We can still theoretically deal with multigraphs like so:

 - simplify the graph (no self-loops / multi-edges)
 - colour edges based on their original multiplicity (simple edges get 1, edges merged from two parallel ones get 2, etc.)
 - colour vertices based on how many incident self-loops they had

Now we can use coloured-graph isomorphism (using VF2) to handle the multigraph.

This PR adds a function to do this graph simplification while counting self-loops and multi-edges.

It also rearranges the Isomorphism documentation sections a bit.

Please verify before merging!  It closely follows `igraph_simplify` in its implementation.

Question: In undirected graphs, do edges always appear sorted, i.e. always 1-2 and not 2-1?  `igraph_simplify` seems to rely on this where it checks `to == pto && from == pfrom` but not `to == pfrom && from == pto`.  I did the same in this function

szhorvat commented:
This should probably be renamed to something better: any suggestions?

I implemented multigraph subgraph matching in IGraph/M based on this. It would be nice to include that in the igraph C core eventually. It's not as trivial as using colours because a smaller multiplicity edge in the pattern graph should match the higher multiplicity edge in the target graph.  Thus we need a custom edge comparator function, which is not available in the high-level interfaces. Even if it were, constant callbacks to the high-level interface would kill performance.

szhorvat commented:
Related (Tamás's answer): 

 - https://stackoverflow.com/questions/33980048/checking-for-multigraph-isomorphism-in-python-igraph

ntamas commented:
>  In undirected graphs, do edges always appear sorted, i.e. always 1-2 and not 2-1?

Right now this is indeed the case.

> This should probably be renamed to something better: any suggestions?

How about `igraph_simplify_and_colorize()`?

Otherwise the implementation seems okay to be, but as always, there should be some tests in `examples/simple` that verify the implementation. The ones for `igraph_simplify()` could probably be re-used.

szhorvat commented:
Since the motivation for adding this function is to handle isomorphism for non-simple graphs, shall we do that at the same time, and merge everything together? Some issues might appear along the way.

If yes, would it be better to extend `igraph_isomorphic` and `igraph_subisomorphic` to handle multigraphs, or create entirely new functions?  I think that extending them are a good idea since these are "generic" functions that dispatch to the "best" implementation for each specific case.

`igraph_isomorphic` and `igraph_subisomorphic` only return a yes/no answer but they do not return an actual mapping. Eventually, it may be a good idea to also add generic functions which return one or more mapping. In particular for the subgraph case, one would typically want to know _where_ a specific subgraph appears.  It's not clear to me what's the best approach here: generic function that returns _one_ mapping, generic function that returns multiple mappings (if requested), or specific functions (i.e. `vf2` in the name) that return multiple mappings.  The main motivation for adding such a function is that it is currently not easily implementable for subgraphs/multigraphs in the high-level interfaces.

For the subgraph case, in principle this is implementable both for VF2 and LAD. I would not bother with LAD for now because later versions of LAD handle colours differently than LADv1 (which is in igraph). If LAD ever gets updated, this would have to change as well.

The state-of-the-art algorithms that could be used for this are: [RI](https://github.com/InfOmics/RI), [VF3](https://github.com/vincarlet/vf3lib/) and [LADv3](https://perso.liris.cnrs.fr/christine.solnon/LAD).  

I experimented a bit with VF3 and the naive integration is often slower than the existing VF2, so I put it aside for now. I suspect that this either has to do with converting the graph format (I did this for `vf3lib` but igraph's VF2 is well-integrated and _maybe_ it doesn't do a conversion), or more likely with the pre-processing step (it's not clear to me when that's actually necessary).

szhorvat commented:
Should the `res` argument come before or after the `graph` argument? igraph is not entirely consistent in this and has functions of both kinds.

szhorvat commented:
>> In undirected graphs, do edges always appear sorted, i.e. always 1-2 and not 2-1?
>
> Right now this is indeed the case.

This should be documented as an invariant, just in case someone contributes a function that messed with the internal graph structure directly ...

On a related note, how do I build the documentation so that it includes `\ingroup internal` things, such as `igraph_t` itself?

ntamas commented:
> If yes, would it be better to extend igraph_isomorphic and igraph_subisomorphic to handle multigraphs, or create entirely new functions? I think that extending them is a good idea since these are "generic" functions that dispatch to the "best" implementation for each specific case.

I would vote for extending `igraph_isomorphic` and `igraph_subisomorphic`.

> Eventually, it may be a good idea to also add generic functions which return one or more mappings.

Yes, indeed.

> Should the `res` argument come before or after the `graph` argument? igraph is not entirely consistent in this and has functions of both kinds.

I know that we haven't been consistent with the API in several cases. Right now I would say that I would follow the convention that the first argument is always the "this" argument (i.e. what would be "this" if we were programming in C++), just like the explicit `self` argument in Python. The second one could be the output argument (if there is one), and the input arguments could come after that.

ntamas commented:
> On a related note, how do I build the documentation so that it includes \ingroup internal things, such as igraph_t itself?

Sorry, I have no idea :( The whole documentation build system was brewed in-house and I don't know exactly how it works by now. I know that a long while ago we started with Doxygen, but then it turned out that it did not really satisfy our needs (we wanted to generate documentation in HTML, PDF and INFO formats), so we wrote a quick regexp-based converter (`doc/doxrox.py`) that parses the C source code and converts everything into Docbook format (this is done by the rules defined in `doc/c-docbook.re`). Some of the Doxygen-related tags (such as `\ingroup`) were not needed in this new setup so we simply instructed `doxrox.py` to ignore these tags. I don't think that the `\ingroup` tags are actually used for anything by now - the documentation's XXML files basically specify explicitly what needs to be included where.

szhorvat commented:
> > Should the `res` argument come before or after the `graph` argument? igraph is not entirely consistent in this and has functions of both kinds.
> 
> I know that we haven't been consistent with the API in several cases. Right now I would say that I would follow the convention that the first argument is always the "this" argument (i.e. what would be "this" if we were programming in C++), just like the explicit `self` argument in Python. The second one could be the output argument (if there is one), and the input arguments could come after that.

That's how I think of it too, but then the question can be rephrased:

Is this function a constructor (that takes a multigraph as input)? Or a method of `igraph_t` that outputs another graph?

It makes sense to say that it's a method that outputs a graph because it _also_ outputs the colours (multiplicities).  Then we keep it as it currently it (input graph is the first, output graph is the second).

ntamas commented:
In the face of ambiguity, I'd say let's keep stuff as it is now :)

szhorvat commented:
I forgot about attribute combiners, but given that this is specifically meant as a helper function to deal with multigraph isomorphism, I think it's fine not to include that. Agreed?

ntamas commented:
Yes.

szhorvat commented:
Added tests. Also added a file with utility functions for testing.  Once merged, I'll send a PR which adds tests for `igraph_tree` (see #1097). I have them but they use the new helper functions.

szhorvat commented:
An explanation for why the vector printing and the graph printing are so fancy in the helper functions:

This is so that it can be clear when a vector is empty or a graph has no edges. It would be confusing if a test failed because the number of empty lines didn't match with the expected output.

ntamas commented:
Seems good to me!
